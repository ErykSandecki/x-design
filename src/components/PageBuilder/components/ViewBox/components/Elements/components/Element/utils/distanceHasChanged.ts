import { RefObject } from 'react';

// types
import { T2DCoordinates } from 'types';

export const distanceHasChanged = (
  event: MouseEvent,
  cursorPosition: RefObject<T2DCoordinates>,
): boolean => {
  const dx = event.clientX - cursorPosition.current.x;
  const dy = event.clientY - cursorPosition.current.y;
  const distance = Math.sqrt(dx * dx + dy * dy);

  return distance > 5;
};
