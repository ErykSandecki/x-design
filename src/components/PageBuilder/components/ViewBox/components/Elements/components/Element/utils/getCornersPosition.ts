// others
import { OUTLINE } from '../constants';

// types
import { TElementDynamicData } from 'store/pageBuilder/types';

export const getCornersPosition = (
  height: TElementDynamicData['height']['value'],
  width: TElementDynamicData['width']['value'],
): TRectCoordinates => ({
  x1: (OUTLINE / 2) * -1,
  x2: (width as number) + OUTLINE / 2,
  y1: (OUTLINE / 2) * -1,
  y2: (height as number) + OUTLINE / 2,
});
