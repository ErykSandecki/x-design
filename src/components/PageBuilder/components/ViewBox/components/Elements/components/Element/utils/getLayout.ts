import { CSSProperties } from 'react';

// types
import { LayoutType, TElement } from 'types';

export const getLayout = (layout: TElement['layout']): CSSProperties => {
  switch (layout.type) {
    case LayoutType.default:
      return {};
    case LayoutType.vertical:
      return {
        display: 'flex',
        flexDirection: 'column',
      };
    case LayoutType.horizontal:
      return {
        display: 'flex',
      };
    default:
      return {
        display: 'grid',
      };
  }
};
