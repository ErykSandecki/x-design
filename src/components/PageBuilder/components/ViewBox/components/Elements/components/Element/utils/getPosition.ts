import { CSSProperties } from 'react';

// types
import { AlignmentHorizontal, AlignmentVertical, TElement } from 'types';

export const getHorizontalPosition = (
  aligment: TElement['alignment'],
  x: T2DCoordinates['x'],
): {
  left?: string;
  originX: string;
  right?: string;
  translateX: string;
} => {
  switch (aligment.horizontal) {
    case AlignmentHorizontal.center:
      return {
        left: '50%',
        originX: '0',
        translateX: '-50%',
      };
    case AlignmentHorizontal.left:
      return {
        left: '0',
        originX: 'center',
        translateX: '0',
      };
    case AlignmentHorizontal.right:
      return {
        originX: 'center',
        right: '0',
        translateX: '0',
      };
    default:
      return {
        left: `${x}px`,
        originX: 'center',
        translateX: '0',
      };
  }
};

export const getVerticalPosition = (
  aligment: TElement['alignment'],
  y: T2DCoordinates['y'],
): {
  bottom?: string;
  originY: string;
  top?: string;
  translateY: string;
} => {
  switch (aligment.vertical) {
    case AlignmentVertical.bottom:
      return {
        bottom: '0',
        originY: 'center',
        translateY: '0',
      };
    case AlignmentVertical.center:
      return {
        originY: '0',
        top: '50%',
        translateY: '-50%',
      };
    case AlignmentVertical.top:
      return {
        originY: 'center',
        top: '0',
        translateY: '0',
      };
    default:
      return {
        originY: 'center',
        top: `${y}px`,
        translateY: '0',
      };
  }
};

export const getPosition = (
  alignment: TElement['alignment'],
  angle: number,
  x: T2DCoordinates['x'],
  y: T2DCoordinates['y'],
): CSSProperties => {
  const { originX, translateX, ...horizontal } = getHorizontalPosition(alignment, x);
  const { originY, translateY, ...vertical } = getVerticalPosition(alignment, y);

  return {
    ...horizontal,
    ...vertical,
    transform: `rotate(${angle}deg) translate(${translateX}, ${translateY})`,
    transformOrigin: `${originX} ${originY}`,
  };
};
