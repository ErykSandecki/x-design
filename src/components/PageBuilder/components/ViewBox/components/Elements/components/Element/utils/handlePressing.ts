import { MouseEvent } from 'react';

export const handlePressing = (event: MouseEvent, setIsPressing: TFunc<[boolean]>): void => {
  if (!event.shiftKey) {
    setIsPressing(true);
  }
};
