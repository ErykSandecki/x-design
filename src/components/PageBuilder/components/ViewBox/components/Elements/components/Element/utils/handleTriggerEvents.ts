import { Dispatch } from 'redux';
import { defer } from 'lodash';

// store
import { changeParent } from 'store/pageBuilder/actions';
import { eventsSelector } from 'store/pageBuilder/selectors';
import { store } from 'store';

export const handleTriggerEvents = (dispatch: Dispatch): void => {
  const { draggableElements, possibleIndexPosition, possibleParent } = eventsSelector(store.getState());

  defer(() => dispatch(changeParent(draggableElements, possibleIndexPosition, possibleParent)));
};
