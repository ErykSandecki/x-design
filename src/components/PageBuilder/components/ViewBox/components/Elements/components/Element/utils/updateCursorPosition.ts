import { RefObject } from 'react';

// types

export const updateCursorPosition = (
  cursorPosition: RefObject<T2DCoordinates>,
  event: MouseEvent | React.MouseEvent,
): void => {
  cursorPosition.current = {
    x: Math.round(event.clientX),
    y: Math.round(event.clientY),
  };
};
