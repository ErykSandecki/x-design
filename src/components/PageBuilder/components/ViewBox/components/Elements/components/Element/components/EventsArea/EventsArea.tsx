import { createPortal } from 'react-dom';
import { FC, memo, RefObject } from 'react';

// components
import Corners from '../../../../../Corners/Corners';
import TransformArea from '../../../../../TransformArea/TransformArea';
import { Box } from 'shared';

// core
import { useRefs } from 'pages/PageBuilderPage/core/RefsProvider';

// types
import { MouseMode, T2DCoordinates, TElement } from 'types';

// utils
import { getCornersPosition } from '../../utils/getCornersPosition';

type TOutlineProps = {
  absoluteCoordinates?: T2DCoordinates;
  elementRef: RefObject<HTMLDivElement>;
  height: TElement['height'];
  id: TElement['id'];
  mouseMode: MouseMode;
  relativeCoordinates?: TElement['coordinates'];
  rotate: TElement['rotate'];
  width: TElement['width'];
};

const EventsArea: FC<TOutlineProps> = ({
  absoluteCoordinates,
  elementRef,
  height,
  id,
  mouseMode,
  relativeCoordinates,
  rotate,
  width,
}) => {
  const rectCoordinates = getCornersPosition(height, width);
  const { overlayContainerRef } = useRefs();

  return createPortal(
    <Box
      style={{
        height,
        left: `${absoluteCoordinates.x}px`,
        top: `${absoluteCoordinates.y}px`,
        transform: `rotate(${rotate}deg)`,
        transformOrigin: 'center center',
        width,
      }}
      sx={{ position: 'absolute' }}
    >
      {/* // Technical note: Box blank to allow move block when overflow */}
      <Box style={{ height, width }} />
      <Corners rectCoordinates={rectCoordinates} />
      <TransformArea
        elementRef={elementRef}
        height={height}
        id={id}
        moseMode={mouseMode}
        rotate={rotate}
        x={relativeCoordinates.x}
        y={relativeCoordinates.y}
        width={width}
      />
    </Box>,
    overlayContainerRef.current,
  );
};

export default memo(EventsArea);
