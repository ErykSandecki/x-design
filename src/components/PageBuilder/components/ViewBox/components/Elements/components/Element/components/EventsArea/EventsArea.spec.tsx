import { Provider } from 'react-redux';
import { RefObject } from 'react';
import { render } from '@testing-library/react';

// components
import EventsArea from './EventsArea';

// core
import { RefsProvider } from 'pages/PageBuilderPage/core/RefsProvider';

// others
import { BASE_2D } from 'shared';

// store
import { configureStore } from 'store/store';

// types
import { MouseMode } from 'types';

// utils
import { createHtmlElement } from 'utils';

const elementRef = {
  current: createHtmlElement('div'),
} as RefObject<HTMLDivElement>;
const overlayContainer = document.createElement('div');

describe('EventsArea snapshots', () => {
  beforeAll(() => {
    // mock
    document.body.appendChild(overlayContainer);
  });

  it('should render EventsArea', () => {
    // mock
    const store = configureStore();

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <RefsProvider overlayContainerRefHtml={overlayContainer}>
          <EventsArea
            absoluteCoordinates={BASE_2D}
            counterAngle={0}
            elementRef={elementRef}
            height={100}
            id="1"
            mouseMode={MouseMode.default}
            relativeCoordinates={BASE_2D}
            rotate={0}
            width={100}
          />
        </RefsProvider>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
