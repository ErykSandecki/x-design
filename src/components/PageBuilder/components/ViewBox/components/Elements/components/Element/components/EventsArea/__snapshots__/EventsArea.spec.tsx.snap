// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsArea snapshots should render EventsArea 1`] = `
<DocumentFragment>
  <div
    class="css-1xpffs9"
    data-test-box=""
    style="height: 100px; left: 0px; top: 0px; transform: rotate(0deg); transform-origin: center center; width: 100px;"
  >
    <div
      class="css-1jnxvbp"
      data-test-box=""
      style="height: 100px; width: 100px;"
    />
    <svg
      class="Corners Corners--light"
      style="height: 102px; left: -1px; top: -1px; width: 102px;"
    >
      <rect
        height="6"
        rx="1.5"
        width="6"
        x="-3"
        y="-3"
      />
      <rect
        height="6"
        rx="1.5"
        width="6"
        x="-3"
        y="97.5"
      />
      <rect
        height="6"
        rx="1.5"
        width="6"
        x="97.5"
        y="-3"
      />
      <rect
        height="6"
        rx="1.5"
        width="6"
        x="97.5"
        y="97.5"
      />
    </svg>
    <div
      class="TransformArea TransformArea--light"
      style="height: 100px; width: 100px;"
    >
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--east TransformArea__anchor-resize--east--light"
        data-test-anchor-resize="east"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--north TransformArea__anchor-resize--north--light"
        data-test-anchor-resize="north"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--northEast TransformArea__anchor-resize--northEast--light"
        data-test-anchor-resize="north-east"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--northWest TransformArea__anchor-resize--northWest--light"
        data-test-anchor-resize="north-west"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--south TransformArea__anchor-resize--south--light"
        data-test-anchor-resize="south"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--southEast TransformArea__anchor-resize--southEast--light"
        data-test-anchor-resize="south-east"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--southWest TransformArea__anchor-resize--southWest--light"
        data-test-anchor-resize="south-west"
      />
      <div
        class="TransformArea__anchor-resize TransformArea__anchor-resize--light TransformArea__anchor-resize--west TransformArea__anchor-resize--west--light"
        data-test-anchor-resize="west"
      />
      <div
        class="TransformArea__anchor-rotate TransformArea__anchor-rotate--light TransformArea__anchor-rotate--northEast TransformArea__anchor-rotate--northEast--light"
        data-test-anchor-rotate="north-east"
      />
      <div
        class="TransformArea__anchor-rotate TransformArea__anchor-rotate--light TransformArea__anchor-rotate--northWest TransformArea__anchor-rotate--northWest--light"
        data-test-anchor-rotate="north-west"
      />
      <div
        class="TransformArea__anchor-rotate TransformArea__anchor-rotate--light TransformArea__anchor-rotate--southEast TransformArea__anchor-rotate--southEast--light"
        data-test-anchor-rotate="south-east"
      />
      <div
        class="TransformArea__anchor-rotate TransformArea__anchor-rotate--light TransformArea__anchor-rotate--southWest TransformArea__anchor-rotate--southWest--light"
        data-test-anchor-rotate="south-west"
      />
    </div>
  </div>
</DocumentFragment>
`;
